packages/salesforce-adapter/src/filters/custom_objects_instances.ts

```diff
@@ -0,0 +1,129 @@
+  const iter = itr[Symbol.asyncIterator]()
+  const next = async (): Promise<void> => {
+    const curr = await iter.next()
+    if (curr.done) return undefined
+    if (_.isArray(curr.value) && curr.value.length > 0) {
+      res.push(...transformer(curr.value))
+    }
+    return next()
+  }
+  await next()
+  return res
+}
+
+const getObjectInstances = async (
+  client: SalesforceClient,
+  object: ObjectType
+): Promise<Array<InstanceElement>> => {
+  const recordsToInstances = (records: salesforceRecord[]): InstanceElement[] => {
+    const recordToInstance = (record: salesforceRecord): InstanceElement => {
+      const getInstancePath = (instanceName: string): string[] => {
+        const objectNamespace = getNamespace(object)
+        if (objectNamespace) {
+          return [SALESFORCE, INSTALLED_PACKAGES_PATH, objectNamespace,
+            RECORDS_PATH, object.elemID.typeName, instanceName]
+        }
+        return [SALESFORCE, RECORDS_PATH, object.elemID.typeName, instanceName]
+      }
+
+      return new InstanceElement(
+        // TODO: Handle elemID with additional logic
+        record.Id,
+        object,
+        record,
+        getInstancePath(record.Id),
+      )
+    }
+
+    // Name sub-fields are returned at top level -> move them to the nameField
+    const transformNameValues = (values: salesforceRecord[]): salesforceRecord[] => {
+      const nameSubFields = Object.keys(Types.compoundDataTypes.Name.fields)
+      // We assume there's only one Name field
+      const nameFieldName = Object.keys(_.pickBy(object.fields, isNameField))[0]
+      return values.map(value => ({
+        ..._.omit(value, nameSubFields),
+        ...nameFieldName !== undefined ? { [nameFieldName]: _.pick(value, nameSubFields) } : {},
+        Id: value.Id,
+      }))
+    }
+
+    const instanceValues = transformNameValues(records)
+    return instanceValues.map(recordToInstance)
+  }
+  const queryString = buildQueryString(object)
+  const recordsIterable = await client.queryAll(queryString)
+  return consumeAsyncIterable(recordsIterable, recordsToInstances)
+}
+
+const getObjectTypesInstances = async (
+  client: SalesforceClient,
+  objects: ObjectType[]
+): Promise<InstanceElement[]> =>
+  (_.flatten(await Promise.all(objects.map(o => getObjectInstances(client, o)))))
+
+const filterObjectTypes = (elements: Element[], namespaces: string[]): ObjectType[] =>
+  (elements
+    .filter(isObjectType)
+    .filter(isCustomObject)
+    .filter(e => {
+      const namespace = getNamespace(e)
+      return namespace !== undefined && namespaces.includes(namespace)
+    }))
+
+const filterCreator: FilterCreator = ({ client, config }) => ({
+  onFetch: async (elements: Element[]) => {
+    if (_.isUndefined(config.namespacesToFetchInstancesFor)
+      || config.namespacesToFetchInstancesFor === []) {
+      return
+    }
+    const relevantObjectTypes = filterObjectTypes(elements, config.namespacesToFetchInstancesFor)
```

Comment:
nit: the current condition assumes something about the implementation of `filterObjectTypes`.
maybe worth changing to something a little less efficient:
```suggestion
    const relevantObjectTypes = filterObjectTypes(elements, config.namespacesToFetchInstancesFor)
    if (relevantObjectTypes.length === 0) {
      return
    }
```